model/model_training/models/peft_modeling.py

```diff
@@ -7,22 +7,42 @@
 from peft import LoraConfig, PeftModel, PrefixTuningConfig, get_peft_model, prepare_model_for_int8_training
 
 
+def add_embeddings(model, embed_path, tokenizer):
+    old_embeddings = model.get_input_embeddings()
+    old_num_tokens, old_embedding_dim = old_embeddings.weight.size()
+    new_embeddings = torch.nn.Embedding(old_num_tokens, old_embedding_dim)
+    new_embeddings.to(old_embeddings.weight.device, dtype=old_embeddings.weight.dtype)
+    model._init_weights(new_embeddings)
+    embed_weights = torch.load(embed_path, map_location=old_embeddings.weight.device)
+    vocab_size = tokenizer.vocab_size
+    new_embeddings.weight.data[:vocab_size, :] = old_embeddings.weight.data[:vocab_size, :]
+    new_embeddings.weight.data[vocab_size : vocab_size + embed_weights.shape[0], :] = embed_weights.to(
+        new_embeddings.weight.dtype
+    ).to(new_embeddings.weight.device)
+    model.set_input_embeddings(new_embeddings)
+    model.tie_weights()
+
+
 def load_peft_model(model, peft_model_path, tokenizer):
-    model.resize_token_embeddings(len(tokenizer))
+    embed_weights = hf_hub_download(peft_model_path, "extra_embeddings.pt")
+    model.resize_token_embeddings(tokenizer.vocab_size + torch.load(embed_weights).shape[0])
     model.config.eos_token_id = tokenizer.eos_token_id
     model.config.bos_token_id = tokenizer.bos_token_id
     model.config.pad_token_id = tokenizer.pad_token_id
     model = PeftModel.from_pretrained(
         model,
-        peft_model_path,
+        model_id=peft_model_path,
         torch_dtype=model.dtype,
     )
     model.eos_token_id = tokenizer.eos_token_id
-    extra_embeds = hf_hub_download(peft_model_path, "extra_embeddings.pt")
-    embed_weights = torch.load(extra_embeds, map_location=model.device)
-    model.base_model.model.model.embed_tokens.weight[len(tokenizer) - embed_weights.shape[0] :, :] = embed_weights.to(
-        model.base_model.model.model.embed_tokens.weight.dtype
-    )
+    add_embeddings(model, embed_weights, tokenizer)
+    return model
+
+
+def load_peft_finetuned_model(model, peft_model_path, tokenizer):
```

Comment:
What is the difference between `save_adapter_model_from_ckpt` and something like

```
class PeftSavingCallback(TrainerCallback):
    def on_train_end(self, args, state, control, **kwargs):
        kwargs["model"].save_pretrained("model/.saved_models/adapter")
```

ie. what is the purpose behind the `extra_embeddings` and where are they added?

I am also a bit confused about where the `save_adapter_model_from_ckpt(cfg)` gets called in trainer_sft  and where `SaveLoraConfig` is initialized.

Reply:
The `extra_embeddings` are the 5 embeddings for the special tokens. They don't exist with the base model and are randomly initialized at the beginning of training. Although they are frozen and don't change from initialization, they need to be saved with the adapter/LoRA to recreate the full model with the base model.
`
save_adapter_model_from_ckpt(cfg)` doesn't get called and is not initialized. Currently the HF Trainer saves a pytorch_model.bin along with the DS ckpt correctly for each stage (offload etc.) So the weights can always be extracted post training.

This code is just to extract the adapter weights and special token embeddings from the pytorch_model.bin and is not integrated anywhere (e.g. as a callback)

This code could be moved into a new script: `tools/export_LoRA_model.py` which then extracts the adapter/embeddings and also has a flag to merge the model (so same as a full fine-tuned set of model weights) and uploads one or both to the hub. 
One simplification to the code is we could have OA base model versions on the hub which are the same as the base model but have extra embeddings, so we don't need the embedding code.

Reply:
@jordiclive Thanks for the detailed response!

Reply:
Hi @jordiclive sorry to necro this thread. I was wondering whether this code is necessary when LoRA finetuning one of the models that has already been trained on oasst dataset like `falcon-7b-sft-mix-2000`.

In this case, wouldn't the extra embeddings be present in the model ?